src/ServiceProviders/CIProviders/GitLabCI/GitLabCIProvider.php

```diff
@@ -0,0 +1,187 @@
+        );
+
+        return [ $gitlabTokenRequest ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setCredentials(CredentialProviderInterface $credentials_provider)
+    {
+        // Since the `credentialRequests()` method declared that we need a
+        // GITLAB_TOKEN credential, it will be available for us to copy from
+        // the credentials provider when this method is called.
+        $tokenKey = self::GITLAB_TOKEN;
+        $token = $credentials_provider->fetch($tokenKey);
+        if (!$token) {
+            throw new \Exception('Could not determine authentication token for GitLab serivces. Please set ' . $tokenKey);
+        }
+        $this->setToken($token);
+    }
+
+    public function projectUrl(CIState $ci_env)
+    {
+        $repositoryAttributes = $ci_env->getState('repository');
+        return 'https://' . $this->GITLAB_URL . '/' . $repositoryAttributes->projectId();
+    }
+
+    protected function apiUrl(CIState $ci_env)
+    {
+        $repositoryAttributes = $ci_env->getState('repository');
+        $apiRepositoryType = $repositoryAttributes->serviceName();
+        $target_project = urlencode($repositoryAttributes->projectId());
+
+        return "https://" . $this->GITLAB_URL . "/api/v4/projects/$target_project/variables";
+    }
+
+    /**
+     * Return the text for the badge for this CI service.
+     */
+    public function badge(CIState $ci_env)
+    {
+        $url = $this->projectUrl($ci_env);
+        return "[![GitLabCI]($url/bradges/master/build.svg?style=shield)]($url)";
+    }
+
+    /**
+     * Write the CI environment variables to the GitLabCI "envrionment variables" configuration section.
+     *
+     * @param CIState $ci_env
+     * @param Session $session TEMPORARY to be removed
+     */
+    public function configureServer(CIState $ci_env)
+    {
+        $this->logger->notice('Configure GitLab CI');
+        $this->setGitLabCIEnvironmentVars($ci_env);
+    }
+
+    protected function setGitLabCIEnvironmentVars(CIState $ci_env)
+    {
+        $gitlab_url = $this->apiUrl($ci_env);
+        $env = $ci_env->getAggregateState();
+        foreach ($env as $key => $value) {
+            $data = ['key' => $key, 'value' => $value];
+            $this->gitlabCIAPI($data, $gitlab_url);
+        }
+    }
+
+    public function startTesting(CIState $ci_env) {
+        // Do nothing...it starts automatically.
+    }
+
+  public function addPrivateKey(CIState $ci_env, $privateKey)
+    {
+        // We need to set the SSH Key variable in GitLabCI
+        $gitlab_url = $this->apiUrl($ci_env);
+        $data = ['key' => 'SSH_PRIVATE_KEY', 'value' => file_get_contents($privateKey)];
+        $this->gitlabCIAPI($data, $gitlab_url);
+    }
+
+    protected function gitlabCIAPI($data, $url, $method = 'GET')
+    {
+        $headers = [
+            'Content-Type' => 'application/json',
+            'User-Agent' => 'pantheon/terminus-build-tools-plugin'
```

Comment:
This constant is getting duplicated in a number of places. There should probably be a `const` for it somewhere that all providers can use.